### PR TITLE
CRM-15371 - Move tag management from admin to tag

### PR DESCRIPTION
**Pull request by f2boot**
CRM-15371 - Adds +manage tags+ permission and move tag management out of admin path/folder to its own (as it is for group) so you don't need admin right to manage tags anymore

---
- [CRM-15371: Allow non-administrative users to see and manage tags without Administer CiviCRM permission](https://issues.civicrm.org/jira/browse/CRM-15371)

### CODE REVIEW

**Comment by colemanw** [comment]
@civicrm-builder add to whitelist

**Review by jitendrapurohit — COMMENTED** [review]
While working on it for QA, I was able to create tags without newly introduced `manage tags` permission. Though navigation menu is not displayed, `civicrm/tag&reset=1` is accessible to a user not holding this permission. Maybe, we need to check for the same while the page/form is loaded.

**Comment by seamuslee001** [comment]
Jenkins re test this please

File: xml/templates/civicrm_navigation.tpl
Comment (seamuslee001) [inline xml/templates/civicrm_navigation.tpl]: @f2boot should the permission here be manage tags?

**Review by seamuslee001 — CHANGES_REQUESTED** [review]
Can you please include the upgrade script to update civicrm_navigation and relevant tables

**Comment by monishdeb** [comment]
Marking it with [wip] as the PR need changes mentioned above.

**Comment by eileenmcnaughton** [comment]
@omarabuhussein Jamie told me I could get you to check this out .....

File: xml/templates/civicrm_navigation.tpl
Comment (f2boot) [inline xml/templates/civicrm_navigation.tpl]: I don't know...
This menu/navigation entry is in the "admin" menu that is, for all the rest, with "administer CiviCRM" permission. Setting it to "manage tags" would not change anything since it would not appear anyway if parent "admin" menu is not displayed (cf "administer CiviCRM" permission). 
There are two logics and I don't know which is preferred by core maintainers:
- "manage tags" as for the other navigation entry pointing to this page (from Contact menu)
- "administer CiviCRM" as all other navigation entry from the admin menu

File: CRM/Core/xml/Menu/Tag.xml
Comment (eileenmcnaughton) [inline CRM/Core/xml/Menu/Tag.xml]: I think this adminGroup line should probably go

**Comment by f2boot** [comment]
Committed for civicrm_navigation upgrade script, 
don't now if civicrm_menu should be upgraded through this script or if upgrade runs "civicrm/menu/rebuild?reset=1"
just in case, the lines below can be added to upgrade script 4.7.17.mysql.tpl for civicrm_menu:
UPDATE civicrm_menu SET
  `path` = 'civicrm/tag',
  `access_arguments` = 'a:2:{i:0;a:1:{i:0;s:11:"manage tags";}i:1;s:3:"and";}',
  `page_callback` = 'CRM_Tag_Page_Tag'
WHERE `path`= 'civicrm/admin/tag';

UPDATE civicrm_menu SET
  `path` = 'civicrm/tag/add',
  `access_arguments` = 'a:2:{i:0;a:1:{i:0;s:11:"manage tags";}i:1;s:3:"and";}',
  `page_callback` = 'CRM_Tag_Page_Tag'
WHERE `path`= 'civicrm/admin/tag/add';

**Review by eileenmcnaughton — COMMENTED** [review]
Overall I think this code is good. Obviously the merge conflict needs resolving :-(

However, my other reservation is that administer CiviCRM will no longer be enough to administer tags - what about an OR.

This is a precedent for that

  <item>
     <path>civicrm/admin/scheduleReminders</path>
     <title>Schedule Reminders</title>[A
     <desc>Schedule Reminders.</desc>
     <page_callback>CRM_Admin_Page_ScheduleReminders</page_callback>
     <access_callback>1</access_callback>
     <access_arguments>administer CiviCRM;edit all events</access_arguments>
     <adminGroup>Communications</adminGroup>
     <icon>admin/small/template.png</icon>
     <weight>40</weight>
  </item>

**Comment by eileenmcnaughton** [comment]
Sorry just re-read & you are right - the upgrade script is not required - it is rebuilt on "civicrm/menu/rebuild?reset=1"

**Comment by f2boot** [comment]
thanks for reviewing
I am new in contributing to CiviCRM
trying to resolve conflict by reverting commit, forward to HEAD (forgot to do it before commit) and commit again...

**Comment by eileenmcnaughton** [comment]
So from a git point of view I would probably on your local do

git checkout HEAD~1

(ie. checkout how it was one commit ago) and then (after checking the commits look right) do git push -f  (-f for force)

**Comment by eileenmcnaughton** [comment]
I think the reverted patch is still not applying.

Try interactive rebasing. To do that do 

git fetch -f origin 
(just to make sure your local is up to date)

git rebase -i origin/master

in the rebase screen you can mark rows as '(s)quash' to squash them into the above row or delete rows so as not to keep them

Afterwards you need to push with git push -f to override what is up there - obviously you need to be a bit careful using -f

**Comment by f2boot** [comment]
I am lost...
All these commit I did not want to include...

I am afraid I don't understand git well enough to contribute yet

So sorry about the mess I have done.

**Comment by litespeedmarc** [comment]
It happens sometimes (happened 2 me 2 days ago).  If it's not a large
number of files and your good losing history,  it is always possible to
make a backup of code changes / new branch / new pull request (& close this
one)

**Comment by f2boot** [comment]
thank you for your kind and encouraging words
I will do that in the next few hours

Any advice on how to link to new PR and close this one properly and respectfully for reviewers are welcome

**Comment by monishdeb** [comment]
@f2boot seems like PR is not clean since there are unwanted commits. Can you please create a new PR by doing cherry-pick. For reference check https://gist.github.com/unbracketed/889c844473bcca1917e2

This is a technique where we always pick commits (only those required) from one branch to your new branch. Also in order to avoid stale PR like this one, always do git pull --rebase upstream master . Am closing this PR now since it gone stale.

**Comment by eileenmcnaughton** [comment]
@f2boot if you open a new pr the just add a comment linking the 2.

**Comment by eileenmcnaughton** [comment]
Just looking at what has happened here. (assuming your civicrm remote is called origin) If you do

```
git pull origin master
```

If will merge the upstream repo into your local

if you do
```
git pull --rebase origin master
```

if will remove your commits, pull from master, and reapply you commits over the top. If there are merge conflicts then it will stop & you can edit the files & add the changes & then do 

``
git rebase --continue
``

If you have changed your git history you can only push it back to your branch using the -f switch

``
git push -f myrepo master
``
(where your repo is named myrepo)

Using the -f switch is not considered good on a repo where you collaborate with others but is fine on your own repo but be a bit more careful when using it

**Comment by f2boot** [comment]
@litespeedmarc , @eileenmcnaughton, @monishdeb and others, thank you so much for your support, really feels like being part of a community ;)

I have two questions about the proper git process and use of 
git pull --rebase origin master

Do I understand well that
1. it should be done from the "forked" branch I use to make the PR (vs master branch) ?
2. it should be done before any new change to the fork (meant for PR) AND just before git push ? 

thank so much again

**Comment by monishdeb** [comment]
@f2boot 
The difference between merge and rebase (after fetching the latest ```master``` branch changes by ```git fetch upstream master``` )  is this :

- When you do ```git pull upstream master``` you are actually going for merge in other words merging your remote's ```origin``` branch ```move_tag_management_from_admin_to_tag``` with remote's ```upstream``` branch ```master```. Now this might result into introducing unwanted conflicts and false merge commits into your branch and eventually got exposed in your PR like here. 

- When you do ```git pull --rebase upstream master``` you are applying the latest changes as commits one by one and then it tries to append your commit at the end  but then there is a chance that your commit may conflict with the latest code-base. In this case we use to manually resolve those conflicts manually in desired files. Then do ```git rebase --continue``` 

But now this PR can't be resolved by rebase so you need to create new branch linked to remote ```master``` which is ``` git checkout upstream/master -b yournewbranch``` Then cherry-pick your desired commits.

**Comment by omarabuhussein** [comment]
hey @eileenmcnaughton  ..  sorry but I was really busy (and still :cry: ), I will try to take a look once I got sometime this week and hopefully before it get merged .

**Comment by eileenmcnaughton** [comment]
@omarabuhussein I think it's OK - @f2boot has written a good patch but just has to work through the git issues

**Comment by f2boot** [comment]
Made new PR as suggested
See #9250

Thanks to all for great support

**Pull request closed without merging**
